File: Android/app/src/main/java/com/pokeapi/lpiem/pokeapiandroid/Provider/PokemonProvider.java
package com.pokeapi.lpiem.pokeapiandroid.Provider;


import android.util.Log;

import com.pokeapi.lpiem.pokeapiandroid.Model.Pokemon.Forms;
import com.pokeapi.lpiem.pokeapiandroid.Model.Pokemon.Pokemon;
import com.pokeapi.lpiem.pokeapiandroid.Model.Pokemon.PokemonAPI;
import com.pokeapi.lpiem.pokeapiandroid.Provider.Pokemon.InterfaceCallBackController;
import com.pokeapi.lpiem.pokeapiandroid.View.MainAppActivity;

import java.util.List;

import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;

public class PokemonProvider implements Callback<Pokemon> {
    private Pokemon pokemonInput;
    private String pokemonName;
    private int nextId = 1;
    private String message;
    private InterfaceCallBackController interfaceCallBackController;

    public void start(final MainAppActivity googleConnexionResult){

        final PokemonAPI pokemonAPI = RetrofitSingleton.getInstance();

        Call<Pokemon> callPokemon = pokemonAPI.getPokemonById(6);

        callPokemon.enqueue(new Callback<Pokemon>() {

            @Override
            public void onResponse(Call<Pokemon> call, Response<Pokemon> response) {
//                Log.d("Reponse","Réponse");
                if(response.isSuccessful()){
                    Pokemon pokemonResulted = response.body();
                    pokemonName = "Nom du pokémon : " + pokemonResulted.getName();
                    setPokemon(pokemonResulted);
                    googleConnexionResult.showPokemon(pokemonResulted);
                }else{
                    Log.d("Erreur", "Erreur de connexion");
                }
            }

            @Override
            public void onFailure(Call<Pokemon> call, Throwable t) {
                Log.e("Erreur",t.getLocalizedMessage());
                t.printStackTrace();
            }
        });
//        Log.d("Nom du pokémon",this.pokemonName+"");
    }

    private void getNextPokemon(){
        /*Call<Pokemon>nextPokemon = this.interfaceAPI.getPokemonById(nextId);
        nextPokemon.enqueue(this);*/
    }

    public void setPokemon(Pokemon poke){
        this.pokemonInput = poke;
    }

    public Pokemon getPokemonInput() {
        return pokemonInput;
    }

    private synchronized void fetchData(Response<Pokemon> response) {
        Pokemon rawPokemon = response.body();

        List<Forms> formList = rawPokemon.getFormsList();
        message += "list people : \n\n";
        // changesList.forEach(people -> System.out.println(people.name));  // lambda expression (enable java 1.8 in project structure  - available only since AP 24...
        Log.d("Nom du pokémon",rawPokemon.getName());
        interfaceCallBackController.onWorkDone(true);
    }


    @Override
    public void onResponse(Call<Pokemon> call, Response<Pokemon> response) {
        if(response.isSuccessful()){
            fetchData(response);
        }else{
            System.out.println("Erreur : "+response.errorBody());
            Log.d("Erreur",response.errorBody()+"");
        }
    }

    @Override
    public void onFailure(Call<Pokemon> call, Throwable t) {
        t.printStackTrace();
    }

    public PokemonProvider(Pokemon result) {
        this.pokemonInput = result;
    }

    public PokemonProvider(){
        this.pokemonInput = new Pokemon();
    }

    public String getPokemonName() {
        return pokemonName;
    }

    public void setPokemonName(String pokemonName) {
        this.pokemonName = pokemonName;
    }
}

File: Android/app/src/main/java/com/pokeapi/lpiem/pokeapiandroid/Provider/Singleton/PokemonProvider.kt
package com.pokeapi.lpiem.pokeapiandroid.Provider.Singleton


import android.util.Log

import com.pokeapi.lpiem.pokeapiandroid.Model.Pokemon.Retrofit.PokemonRetrofit
import com.pokeapi.lpiem.pokeapiandroid.Provider.Pokemon.InterfaceCallBackController
import com.pokeapi.lpiem.pokeapiandroid.View.MainAppActivity

import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response

class PokemonProvider : Callback<PokemonRetrofit> {
    var pokemonRetrofitInput: PokemonRetrofit? = null
    var pokemonName: String? = null
    private var message: String? = null

    fun start(googleConnexionResult: MainAppActivity) {

        val pokemonAPI = RetrofitSingleton.getInstance()

        val callPokemon = pokemonAPI!!.getPokemonById(6)

        callPokemon.enqueue(object : Callback<PokemonRetrofit> {

            override fun onResponse(call: Call<PokemonRetrofit>, response: Response<PokemonRetrofit>) {
                if (response.isSuccessful) {
                    val pokemonRetrofitResulted = response.body()
                    pokemonName = "Nom du pokémon : " + pokemonRetrofitResulted!!.name!!
                    setPokemon(pokemonRetrofitResulted)
                    googleConnexionResult.addPokemonToList(pokemonRetrofitResulted)
                } else {
                    Log.d("Erreur", "Erreur de connexion")
                }
            }

            override fun onFailure(call: Call<PokemonRetrofit>, t: Throwable) {
                Log.e("Erreur", t.localizedMessage)
                t.printStackTrace()
            }
        })
        //        Log.d("Nom du pokémon",this.pokemonName+"");
    }

    fun setPokemon(poke: PokemonRetrofit?) {
        this.pokemonRetrofitInput = poke
    }

    @Synchronized
    private fun fetchData(response: Response<PokemonRetrofit>) {
        val rawPokemonRetrofit = response.body()

        message += "pokemonList people : \n\n"
        // changesList.forEach(people -> System.out.println(people.name));  // lambda expression (enable java 1.8 in project structure  - available only since AP 24...
        Log.d("Nom du pokémon", rawPokemonRetrofit!!.name)
    }


    override fun onResponse(call: Call<PokemonRetrofit>, response: Response<PokemonRetrofit>) {
        if (response.isSuccessful) {
            fetchData(response)
        } else {
            println("Erreur : " + response.errorBody()!!)
            Log.d("Erreur", response.errorBody()!!.toString() + "")
        }
    }

    override fun onFailure(call: Call<PokemonRetrofit>, t: Throwable) {
        t.printStackTrace()
    }
}

